### PR TITLE
pyscripts fixes & updates for pwsh pipeline

### PR DESCRIPTION
**Pull request by gregga**

### CODE REVIEW

File: vsts/templates/pre-test-steps.yaml [hunk 1]
```diff
@@ -3,8 +3,8 @@ parameters:
   repo_address: $(IOTHUB-E2E-REPO-ADDRESS)
   repo_user: $(IOTHUB-E2E-REPO-USER)
   repo_password: $(IOTHUB-E2E-REPO-PASSWORD)
-  image_edgeHub: mcr.microsoft.com/azureiotedge-hub:1.0.7
-  image_edgeAgent: mcr.microsoft.com/azureiotedge-agent:1.0.7
+  image_edgeHub: mcr.microsoft.com/azureiotedge-hub:1.0.6
+  image_edgeAgent: mcr.microsoft.com/azureiotedge-agent:1.0.6
```
Comment (BertKleewein) [inline vsts/templates/pre-test-steps.yaml]: please remove all the changes under vsts/ from this PR.  It looks like your revert changed the files to some random version that isn't the version in master.

Comment (gregga) [reply vsts/templates/pre-test-steps.yaml]: updated from Master

File: pyscripts/inject_into_junit.py [hunk 2]
```diff
@@ -105,4 +105,4 @@ def get_suite_lines_from_log(self, log_lines, suite_name):
         return lines_for_junit
 
 if __name__ == "__main__":
-    junit_processor = InjectIntoJunit(sys.argv[1:])
```
Comment (BertKleewein) [inline pyscripts/inject_into_junit.py]: This file doesn't need to be part of this change

File: pyscripts/remove_edgehub_device.py [hunk 3]
```diff
@@ -48,3 +48,4 @@ def remove_edgehub_device():
 
 if __name__ == "__main__":
     remove_edgehub_device()
+
```
Comment (BertKleewein) [inline pyscripts/remove_edgehub_device.py]: this file doesn't need to be part of this change

File: pyscripts/docker_tags.py [hunk 4]
```diff
@@ -212,3 +212,4 @@ def get_docker_tags_from_commit(language, repo, commit, variant):
     else:
         tags.image_tags.insert(0, "latest")
     return tags
+
```
Comment (BertKleewein) [inline pyscripts/docker_tags.py]: This file doesn't need to be part of this change

File: pyscripts/build_docker_image.py [hunk 5]
```diff
@@ -10,193 +10,237 @@
...
-                if obj["status"] not in [
-                    "Waiting",
-                    "Downloading",
-                    "Verifying Checksum",
-                    "Extracting",
-                    "Preparing",
-                    "Pushing",
-                ]:
-                    print("{}: {}".format(obj["status"], obj["id"]))
+class BuildDockerImage:
```
Comment (BertKleewein) [inline pyscripts/build_docker_image.py]: Was there a reason you put this into a class?  Inserting the spaces makes the delta almost impossible to read because (almost) every single line in the file changed.

File: pyscripts/build_docker_image.py [hunk 6]
```diff
@@ -10,193 +10,237 @@
...
+            print("BUILDING IMAGE")
+            print(print_separator)
+
+            force_flag = 0
+
+            if sys.platform == 'win32':
+                base_url = "tcp://127.0.0.1:2375"
+            else:
+                base_url = "unix://var/run/docker.sock"
+            api_client = docker.APIClient(base_url=base_url)
```
Comment (BertKleewein) [inline pyscripts/build_docker_image.py]: these 5 lines are repeated over and over (setting base_url and creating the api_client).  Can you make these global so there's not so much copy/paste.  (at least set base_url once, but preferably create api_client only once. )

File: pyscripts/build_docker_image.py [hunk 7]
```diff
@@ -10,193 +10,237 @@
...
+
+
+        def get_dockerfile_directory(tags):
+            script_dir = os.path.dirname(os.path.realpath(__file__))
+            return os.path.normpath(os.path.join(script_dir, "../ci-wrappers/" + tags.language))
+
+
+        def print_filtered_docker_line(lineobj):
+            lineobj = lineobj.splitlines()
+            for line in lineobj:
```
Comment (BertKleewein) [inline pyscripts/build_docker_image.py]: This function is a lot to digest.  It's very hard to understand what's going on in here.
lineobj is a string (despite being called an obj), and then after you split it, it becomes an array of strings (still called an obj).
Then you split the array into individual lines.  
For each line, you strip it and decode it.  What does this mean?  the string object has no function called "decode", so I assume that linecode is a buffer instead of a string, but I'm not sure how or why.  I'm also not sure what a "line code" is.
Then, you either strip off an ending "\r\n", or you strip off the last character (if rfind("\r\n" returns -1).  and add a close bracket on the end.  Since you're using splitlines to break up the lines, the string probably doesn't have "\r\n" in it, so it looks like you're removing the last character from the line in a rather roundabout way and then then adding a close bracket.  This confuses me.

File: pyscripts/build_docker_image.py [hunk 8]
```diff
@@ -10,193 +10,237 @@
...
+                                pass
+                        else:
+                            print(obj["status"])
+                    elif "error" in obj:
+                        print ("docker error: {}".format(line))
+                        raise Exception(obj["error"])
+                    elif "Step" in obj or "---" in obj:
+                        print("{}".format(obj).strip())
+                    elif obj == "\n":
+                        pass
```
Comment (BertKleewein) [inline pyscripts/build_docker_image.py]: Wait, is "obj" a string or an object?  With a name like "obj", I assume it's an object.  Above, you're treating it like an object, but here it looks like you're treating it like a string, but print("{}".format(obj).strip()).  makes it look like it's not a string (because you're using the roundabout "{}".format() to convert it to a string, but then you strip it, so there must be leading or trailing spaces, so it must be a string.  Help, what is it?

File: pyscripts/build_docker_image.py [hunk 9]
```diff
@@ -10,193 +10,237 @@
...
+                    else:
+                        lineout = "".format("{}", line)
+                        print(''.join([chr(i) if i < 128 else '\x23' for i in lineout]))
```
Comment (BertKleewein) [inline pyscripts/build_docker_image.py]: Again, converting the line (maybe a string or maybe an object) into a string, and stripping out anything that isn't 7-bit ascii.  Why?  Also, out of curiosity, why do you use 'x23' instead of '#'?

File: pyscripts/build_docker_image.py [hunk 10]
```diff
@@ -10,193 +10,237 @@
...
-        + "./deploy-test-containers.sh --{} {}:{}".format(
-            tags.language, tags.docker_full_image_name, tags.image_tags[0]
+                    try:
+                        sys.stdout.write(json.loads(line.decode("utf-8"))["stream"])
+                    except:
+                        try:
+                            print_filtered_docker_line(line)
+                        except:
+                            out_line = ''.format("{}",line)
+                            print(''.join([i if ord(i) < 128 else '#' for i in out_line]))
```
Comment (BertKleewein) [inline pyscripts/build_docker_image.py]: This only adds to my confusion about print_filtered_docker_line.  You have some try-catches here, some inside that function.  Stripping out everything beyond 7-bit ascii in a third place.  I have no idea what is going on in here.

File: pyscripts/build_docker_image.py [hunk 11]
```diff
@@ -10,193 +10,237 @@
...
+            if sys.platform == 'win32':
+                base_url = "tcp://127.0.0.1:2375"
+            else:
+                base_url = "unix://var/run/docker.sock"
+            api_client = docker.APIClient(base_url=base_url)
+
+            for image_tag in tags.image_tags:
+                print("Pushing {}:{}".format(tags.docker_full_image_name, image_tag))
+                for line in api_client.push(
+                    tags.docker_full_image_name, image_tag.lower(), stream=True, auth_config=auth_config
```
Comment (BertKleewein) [inline pyscripts/build_docker_image.py]: >lower [](start = 59, length = 5)

Why did you need to add the "lower" call here?  I'm guessing that someone somewhere doesn't like mixed case, but who and why?  (if we're making bad tags, we should change docker_tags.py to not make bad tags).

File: pyscripts/build_docker_image.py [hunk 12]
```diff
@@ -10,193 +10,237 @@
...
-        Fore.GREEN
-        + "./deploy-test-containers.sh --{} {}:{}".format(
-            tags.language, tags.docker_full_image_name, tags.image_tags[0]
+                    try:
+                        sys.stdout.write(json.loads(line.decode("utf-8"))["stream"])
+                    except:
+                        try:
+                            print_filtered_docker_line(line)
+                        except:
+                            out_line = ''.format("{}",line)
```
Comment (BertKleewein) [inline pyscripts/build_docker_image.py]: >out_line = ''.format("{}",line) [](start = 28, length = 31)

Is this some sort of magic or a typo?  I _think_ you mean "{}".format(line), or do you mean something else?  also, why not just use str(line) (here and everywhere)?

File: pyscripts/build_docker_image.py [hunk 13]
```diff
@@ -10,193 +10,237 @@
...
+        def print_filtered_docker_line(lineobj):
+            lineobj = lineobj.splitlines()
+            for line in lineobj:
+                try:
+                    linecode = line.strip().decode("utf-8")
+                    linecode = linecode[:linecode.rfind("\r\n")] + '}'
+                    jline = json.loads(linecode)
+                    obj = jline["stream"]
+                except:
+                    print(''.join([chr(i) if i < 128 else '\x23' for i in line]))
```
Comment (BertKleewein) [inline pyscripts/build_docker_image.py]: >print(''.join([chr(i) if i < 128 else '\x23' for i in line])) [](start = 20, length = 61)

reading this, I'm still not sure if line is a string or a buffer.  It looks like it's a buffer here, and it looks like you're converting to 7-bit ascii and replacing everything that isn't 7-bit ascii with a '#'.  Huh?  Why?

Comment (gregga) [reply pyscripts/build_docker_image.py]: Bert - I can totally see the errors and confusion here.  This as an attempt to fix the invalid json format returned by docker.APIClient in some cases - the json that was returned by docker.APIClient was not valid json and passing it to json.loads() threw an exception. The json was missing the closing curley.

Yes, I got a little overzealous with the cut and paste from the stackoverflow "solution", and yes it is double decoded into Ascii7.
I will revisit and fix this code. I think the best solution here is to just catch the exception from json.loads() and print the Ascii7 rather than trying to fix it.

File: pyscripts/build_docker_image.py (hunk 11 above)
Comment (gregga) [reply pyscripts/build_docker_image.py]: In the very first run, BEFORE image is in cache:
tag-image FAILS with error 500 due to the PascalCase used in get_commit_name() in docker_tags.py (line:63)
This is passed to image_tag api_client.tag(tags.docker_image_name, tags.docker_full_image_name, image_tag)
api_client.tag ONLY accepts tag in lower case.
Workaround was to add .lower() to image_tag in tag-image AND push-image in build_docker_image.py
Once the image is tagged with lower case, next run can continue using PascalCase because image is in cache and get-image is not case sensitive.
I can add the .lower() in docker_tags.py:get_commit_name(). This may be a better solution.

File: pyscripts/build_docker_image.py (hunk 13 above)
Comment (gregga) [reply pyscripts/build_docker_image.py]: The best solution here may be to catch the exception from json.loads() and print the invalid json as Ascii7 rather than trying to fix the json.

File: pyscripts/build_docker_image.py (hunk 6 above)
Comment (gregga) [reply pyscripts/build_docker_image.py]: Fixed. made base_url global

File: pyscripts/build_docker_image.py (hunk 7 above)
Comment (gregga) [reply pyscripts/build_docker_image.py]: fixed. handle json exception and print ascii7

File: pyscripts/build_docker_image.py (hunk 8 above)
Comment (gregga) [reply pyscripts/build_docker_image.py]: fixed. handle json exception and print ascii7

File: pyscripts/build_docker_image.py (hunk 9 above)
Comment (gregga) [reply pyscripts/build_docker_image.py]: needed to line.decode() for conversion to ascii7.

File: pyscripts/build_docker_image.py (hunk 10 above)
Comment (gregga) [reply pyscripts/build_docker_image.py]: fixed

File: pyscripts/build_docker_image.py (hunk 12 above)
Comment (gregga) [reply pyscripts/build_docker_image.py]: fixed with line.decode()

File: pyscripts/build_docker_image.py (hunk 11 above)
Comment (BertKleewein) [reply pyscripts/build_docker_image.py]: please pull out the .lower changes. I don't feel comfortable with them.

---
In reply to: [295624635](https://github.com/Azure/iot-sdks-e2e-fx/pull/79#discussion_r295624635) [](ancestors = 295624635)

File: pyscripts/build_docker_image.py [hunk 14]
```diff
@@ -121,34 +130,37 @@ def build_image(tags):
         cache_from=cache_from,
         dockerfile=dockerfile,
     ):
-        try:
-            sys.stdout.write(json.loads(line.decode("utf-8"))["stream"])
-        except KeyError:
```
Comment (BertKleewein) [inline pyscripts/build_docker_image.py]: Can you please add this back in?  I'm not sure why this was removed.

Comment (gregga) [reply pyscripts/build_docker_image.py]: Fixed

File: pyscripts/build_docker_image.py (hunk 11 above)
Comment (gregga) [reply pyscripts/build_docker_image.py]: Done.
